netlink/uevent.go

```diff
@@ -81,19 +85,35 @@ func (e UEvent) Equal(e2 UEvent) (bool, error) {
 	return true, nil
 }
 
+// Parse udev event created by udevd.
+// The format of the data header is internal to udev and defined in libudev-monitor.c - see the udev_monitor_netlink_header struct.
+// go-udev only looks at the "magic" number to filter out possibly invalid packets, and at the payload offset. Other fields of the header
+// are ignored.
+// Note, only some of the fields of the header use network byte order, for the rest udev uses native byte order of the platform.
 func parseUdevEvent(raw []byte) (e *UEvent, err error) {
-	fields := bytes.Split(raw[40:], []byte{0x00}) // 0x00 = end of string
+	// the magic number is stored in network byte order.
+	magic := binary.BigEndian.Uint32(raw[8:])
+	if magic != libudevMagic {
+		return nil, fmt.Errorf("cannot parse libudev event: magic number mismatch")
+	}
+
+	// the payload offset int is stored in native byte order.
+	payloadoff := *(*uint32)(unsafe.Pointer(&raw[16]))
```

Comment:
I'm not comfortable with the "unsafe" package. Could you explain me the reason of this usage ?
Because `raw[16]` seems to have the same value the `payloadOff`, ie : https://play.golang.org/p/SLCMCVxGsSl

Reply:
We need to interpret all 4 bytes (the size of uint32) as the offset, thus the unsafe pointer cast. I don't know if there is any better way of doing this in Go. I'm fine with changing it if there is a better way.

The example you gave only works because the reamining 3 bytes are zeros - you're just taking the first byte (the least-significant). See: https://play.golang.org/p/r_M_shUoohW

Reply:
The `unsafe.Pointer(&raw[16])` expression takes *the address* of first byte and it's re-interpreted as `uint32` in native format of the platform - it's not just the 16th byte. See the modified example https://play.golang.org/p/7bWUVkDJ4WQ where I've changed all 4 bytes of raw to non-zero, so it's more apparent.

The difference between unsafe cast and the two variants you gave (with byte shifting and binary.LittleEndian.Uint32) is in support for endianess on other platforms. They will only work on little endian systems, and they both are equvalents. The unsafe cast will work on little- and big- endian systems.

Reply:
To give you some more perspective (because I've experimented a bit with on when working on this PR), the most elegant way would be something like:
```
if weAreRunningOnLittleEndian {
  binary.LittleEndian.Uint32(raw[16:20]
} else // running on big endian {
  binary.BigEndian.Uint32(raw[16:20]
}
```
The problem with this is finding out whether we're running on big- or little- endian system  :) - I couldn't find any nice method in Go libraries, and googling for a solution suggested some runtime checks that revolved around writing a test value to memory and inspecting it with `unsafe.Pointer`....

Reply:
Ok it is clear for me now, I will use this method now, thanks theses explanations.

Reply:
This is ensured by the `if len(raw) > 40 && bytes.Compare(raw[:8], []byte("libudev\x00")) == 0 {` check in line 142.
Regarding your second question, the magic number uses network byte order (yay for consistency ;)) - see respective comments in lines 92 and 95.

Reply:
Ok for me, sorry repetition, I have no additional question and ready to merge ! :)

Reply:
No problem, thanks!